### DISCUSSION TITLE
Linking osg plugins

### DISCUSSION THREAD

**Discussion by Harsh-patel00** [openscenegraph/OpenSceneGraph#1205] [Q&A] [answered]
Hello,

I'm following the "OpenSceneGraph Beginner's Guide" book mentioned on the osg website.
While trying to run the following code:

 ```
#include <osgViewer/Viewer>
#include <osgDB/ReadFile>

int main()
{
    osgViewer::Viewer viewer;
    viewer.setSceneData(osgDB::readNodeFile("cessna.osg"));

    return viewer.run();
}
```

I'm getting error: ```Error reading file cessna.osg: read error (Could not find plugin to read objects from file "cessna.osg".)```

The code runs and **outputs** an empty blue fullscreen window. But the model shown in the book is not shown.

I'm using the following environment:

vcpkg dry run result: 

```
E:\vcpkg> .\vcpkg.exe install osg:x64-windows --dry-run
Computing installation plan...
The following packages are already installed: osg[core,default-features,examples,fontconfig,freetype,nvtt,openexr,plugins]:x64-windows -> 3.6.5#18
```

IDE: CLion
Compiler: MSVC (Visual Studio 2022)
Compiler architecture: amd64
CMake Build Type: Debug
CMake file: [CMakeLists.txt](https://github.com/openscenegraph/OpenSceneGraph/files/10712812/CMakeLists.txt)
Env. Variables Set: 

- OSG_ROOT: ```E:\vcpkg\packages\osg_x64-windows```
- OSG_LIBRARIES: ```E:\vcpkg\packages\osg_x64-windows\lib```
- OSG_FILE_PATH: ```E:\vcpkg\packages\osg_x64-windows\plugins\osgPlugins```; ```E:\OpenSceneGraph-Data```

OpenSceneGraph Data: [OpenSceneGraph-Data-3.4.0.zip](https://www.openscenegraph.com/downloads/stable_releases/OpenSceneGraph-3.4.0/data/OpenSceneGraph-Data-3.4.0.zip)

I've searched for this issue and everyone says that adding osgPlugin folder to env variables, I tried that but it doesn't work for me (Maybe I'm doing something wrong?), please help!

**Comment by robertosfield** [comment, marked as answer]
THe OSG plugins should normally be placed in a subdirectory below the library directory so just setting PATH variable to that library directory should be enough,

The OSG_FILE_PATH is used for data files.

**Reply by Harsh-patel00** [reply]
I tried your solution, but it didn't work, still has the same error. 
The Plugins folder contains all .dll files, and the lib directory which is in the path contains .lib files.
Also, I removed the plugin's path from the OSG_FILE_PATH.

**Comment by robertosfield** [comment]
On Sat, 11 Feb 2023 at 11:50, Harsh Patel ***@***.***> wrote:

> I tried your solution, but it didn't work, still has the same error.
> The Plugins folder contains all .dll files, and the lib directory which is
> in the path contains .lib files.
> Also, I removed the plugin's path from the OSG_FILE_PATH.
>

The OSG has a notification system that you can alter the verbosity of which
can help spot what paths the plugins are being search for.

set OSG_NOTIFY_LEVEL=DEBUG

Then run the application and look through the output.  You may want to
write the output to a file as it'll be quite verbose.  When looking through
you see the paths the OSG is checking and this might give you a clue to
what is going wrong.

--

As a general comment, the OSG project, like OpenGL that it's based on, is
now in maintenance mode and my expectation is it'll primarily be used for
existing applications that already use the OSG.

For new developers working on new applications then it would probably be
better to look at the VulkanSceneGraph, while it's still early in it's life
so doesn't yet have books written about it, it's much more modern and
forward looking.

**Reply by Harsh-patel00** [reply]
Thank you so much, I've been struggling with this for a while! 
I used ```osgDB::FilePathList pathList = osgDB::Registry::instance()->getLibraryFilePathList();``` for getting the library paths, which osg looks into, for .dll, it did contain the lib directory, but then as you mentioned earlier the plugins folder was not present as a subdirectory. So I moved the plugins folder into the lib directory and the same for the debug lib & plugins folder. 

Now I can run the example! Marking your 1st answer as the answer. Thank you @robertosfield!

**Reply by robertosfield** [reply]
This sounds like it might be a problem with the vcpkg set up.  I don't personally have anything to do with vcpkg so can't comment on how manages or how it be fixed.
